File: SHTFILE.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;

namespace sht {
    class SHTFILE {

        public SHTFILE() { 
        }

        public void init() {
            public playerArg args;
        public List<position> posotions = new List<position>();
        public List<offsets> posotions = new List<offsets>();
        public List<eachShot> shots = new List<eachShot>();


    }

    #region structs
    struct position {
            public float x;
            public float y;
        };

        struct eachShot {
            public byte rate;          //发弹间隔
            public byte delay;         //发弹时间点
            public short power;         //每发子弹的威力
            public position pos;       //发弹点相对坐标
            public position hitBox;    //子弹判定的长和宽
            public float angle;        //发射方向
            public float speed;        //发射速度
            public Int32 unknown1;     //fsl之后加入
            public byte option;        //子机号,0=main
            public byte unknown2;
            public Int16 ANM;           //自机弹贴图
            public Int16 SE;            //音效
            public byte rate2;         //rate2
            public byte delay2;            //delay2
            [MarshalAs(UnmanagedType.ByValArray,SizeConst = 12,ArraySubType = UnmanagedType.I4)]
            public Int32[] flags;
        }
        struct playerArg {
            public Int16 maxPower;                                  //默认4
            public Int16 totalOffset;                               //最大offset数量
            public float hitBox;                                   //判定				(无用)
            public float powerAttractSpeed;                        //收p点的速度		(无用)
            public float powerAttractHitBox;                       //收p点的判定(高速)	(无用)
            public float normalSpeed;                              //高速 速度
            public float forcedSpeed;                              //低速速度
            public float normalDiagonalSpeed;                      //斜向高速(指分配到xy的速度)
            public float forcedDiagonalSpeed;                      //斜向低速
            public Int16 maxPowerLevel;                         //最大的子机数(power的最大百位值)	
            public Int16 unknown1;                                  //不知道
            public Int32 powerVar;                                 //power要修改的内容(40)
            public Int32 maxDamage;                                //单帧子弹最高伤害
            [MarshalAs(UnmanagedType.ByValArray,SizeConst = 5,ArraySubType = UnmanagedType.I4)]
            public Int32[] unknown;                               //不知道
        }
        #endregion
    }
}

File: mainCode/src/com/meng/sht/ShtFile.java
package com.meng.sht;

import java.io.File;
import java.io.FileInputStream;
import java.util.ArrayList;

import com.badlogic.gdx.math.Vector2;
import com.meng.TaiHunDanmaku.ui.GameMain;

public class ShtFile {
	public PlayerArg playerArg = new PlayerArg();
	public ArrayList<Vector2> subPlanePositions = new ArrayList<>();
	public ArrayList<Integer> dataOffsets = new ArrayList<>();
	public ArrayList<Shooter[]> shootersList = new ArrayList<>();
	

    private String name;

    private byte[] fileByte;
    private int position = 0;
    private int shotStart;

	@Override
	public String toString() {
		StringBuilder stringBuilder = new StringBuilder();
		stringBuilder.append("playerArg ").append(playerArg.toString()).append("\n");
		for (Vector2 v : subPlanePositions) {
			stringBuilder.append(v).append(" ");
		}
		stringBuilder.append("\ndataOffsets ");
		for (int i : dataOffsets) {
			stringBuilder.append(i).append(" ");
		}
		stringBuilder.append("\n");
		for (Shooter[] shooters : shootersList) {
			for (Shooter s : shooters) {
				stringBuilder.append(s.toString()).append(" ");
				stringBuilder.append("\n");
			}
		}
		return stringBuilder.toString();
	}
    
    public ShtFile(String fileName) {
    	 name = fileName;
         File ecl = new File(GameMain.baseShtPath + fileName);
         fileByte = new byte[(int) ecl.length()];
         try {
             FileInputStream fin = new FileInputStream(ecl);
             fin.read(fileByte);
         } catch (Exception e) {
             throw new RuntimeException(e.toString());
         }
         playerArg.maxPower=readShort();
         playerArg.totalOffset=readShort();
         playerArg.hitBox=readFloat();
         playerArg.powerAttractSpeed=readFloat();
         playerArg.powerAttractHitBox=readFloat();
         playerArg.normalSpeed=readFloat();
         playerArg.forcedSpeed=readFloat();
         playerArg.normalDiagonalSpeed=readFloat();
         playerArg.forcedDiagonalSpeed=readFloat();
         playerArg.maxPowerLevel=readShort();
         playerArg.unknown1=readShort();
         playerArg.powerVar=readInt();
         playerArg.maxDamage=readInt();
         readIntArray(playerArg.unknown);
         for(int status=0;status<2;++status){ //normal and slow
        	 for(int powerFlag=0;powerFlag<playerArg.maxPower;powerFlag++){ //power 1 to max
        		 for(int shooterFlag=-1;shooterFlag<powerFlag;++shooterFlag){ //power is shooter count
        			 subPlanePositions.add(new Vector2(readFloat(),readFloat()));
        		 }
        	 }
         }
         for (int i = 0; i < playerArg.totalOffset; ++i){ 
        	 dataOffsets.add(readInt());
         }
         shotStart=position;

         for(int i=0;i<dataOffsets.size();++i){
        	 Shooter[] shooters;
        	 int sizeSub1=dataOffsets.size()-1;
        	 if(i<sizeSub1){
        		 shooters =new Shooter[(dataOffsets.get(i+1)-dataOffsets.get(i))/Shooter.size];
        		 } else { 
            	 shooters=new Shooter[(fileByte.length-shotStart-dataOffsets.get(i))/Shooter.size];
            	 }
        	 for(int shooterFlag=0;shooterFlag<shooters.length;++shooterFlag){
        		 Shooter shooter=new Shooter();
        		 shooter.rate=readByte();
        		 shooter.delay=readByte();
        		 shooter.power=readShort();
        		 shooter.pos=new Vector2(readFloat(),readFloat());
        		 shooter.hitBox=new Vector2(readFloat(), readFloat());
        		 shooter.angle=readFloat();
        		 shooter.speed=readFloat();
        		 shooter.unknown1=readInt();
        		 shooter.option=readByte();
        		 shooter.unknown2=readByte();
        		 shooter.ANM=readShort();
        		 shooter.SE=readShort();
        		 shooter.rate2=readByte();
        		 shooter.delay2=readByte();
        		 readIntArray(shooter.flags);
        		 shooters[shooterFlag]=shooter;
        	 }
        	 if(i<sizeSub1){
        		 int split=readInt();
        	 }
        	 shootersList.add(shooters);
         }
    
     
	}

    private byte readByte() {
        return fileByte[position++];
    }
    
    private short readShort() {
        return (short) (fileByte[position++] & 0xff | (fileByte[position++] & 0xff) << 8);
    }

    private int readInt() {
        return (fileByte[position++] & 0xff) | (fileByte[position++] & 0xff) << 8 | (fileByte[position++] & 0xff) << 16 | (fileByte[position++] & 0xff) << 24;
    }
    
    private float readFloat(){
    	return Float.intBitsToFloat(readInt());
    }
    
    private void readIntArray(int[] arr){
    	for (int i = 0; i < arr.length; ++i) {
			arr[i]=readInt();
		}
    }

}